source/code/scxsystemlib/disk/scxlvmutils.cpp

```diff
@@ -229,56 +261,83 @@ namespace SCXSystemLib {
         SCXCoreLib::SCXLogHandle log = SCXCoreLib::SCXLogHandleFactory::GetLogHandle(L"scx.core.common.pal.system.disk.scxlvmutils");
         std::wstringstream       out;
 
-        // First look in Sysfs to find out what partitions the dm device
-        // maps onto.  This is done by listing the slaves entries for the
-        // given dm device.
-        //
-        // Note: The entries in slaves are links and it is the link name
-        //       that is important for this, *not* the resolved link name.
-        std::wstringstream sysfsPath;
-
-        sysfsPath << L"/sys/block/" << ((SCXCoreLib::SCXFilePath) dmDevice).GetFilename() << L"/slaves/";
-
+        // At times, the entries in '/sys/block/<dm-device>/slaves/' point to another dm-device.
+        // in which case, we must navigate to the slaves of that dm device.
+        // we use a non-recursive depth-first traversal.
+        std::stack<std::wstring> dmDeviceStack;
+        dmDeviceStack.push(((SCXCoreLib::SCXFilePath) dmDevice).GetFilename());
         std::vector< SCXCoreLib::SCXFilePath > slaves;
-        try
-        {
-            slaves = m_extDepends->GetFileSystemEntries(sysfsPath.str());
-        }
-        catch (SCXCoreLib::SCXException & e)
+        unsigned int loopCount = 0;
+
+        while(!dmDeviceStack.empty())
         {
+            loopCount++;
+            try
+            {
+                std::wstring currentDeviceName = dmDeviceStack.top();
+                dmDeviceStack.pop();
+                std::vector< SCXCoreLib::SCXFilePath > slaveEntries = m_extDepends->GetFileSystemEntries(GetSysfsPath(currentDeviceName));
+                for( std::vector< SCXCoreLib::SCXFilePath >::const_iterator iter = slaveEntries.begin(); iter != slaveEntries.end(); ++iter)
+                {
+                    if(!PushIfDMSlave(*iter, dmDeviceStack))
+                    {
+                        slaves.push_back(*iter);
+                    }
+                }
+                if (loopCount > maxLoopCount)
+                {
+                   // We assume that the links form an acyclic graph.
+                   // But we still need to be wary of infinite loops - perhaps we are on a corrupted file system, or, for some other reason
+                   // we are in a circular graph.
+                   //
+                   // we bail out if the loop count reaches maxLoopCount.
+                   //
+                   // Note: It is assumed that a cycle of links  will be a very rare scenario. Hence, we are using this "low-tech" way of identifying
+                   // cycles instead of relying on more traditional approaches.
+                   // This way the vast majority of positive scenarios will only pay the price of one integer increment and one comparison.
+                   out.str(L"");
+                   out << L"Crossed " << maxLoopCount << " while navigating links.";
```

Comment:
Did you add this "just in case", or did you hit a case where this actually occurred?

I'm wondering if this should be a "log once" type thing (suspend multiple identical error messages)? And the error message leaves something to be desired. Sure, I know the threshold was exceeded, but how about something like:

L"Exceeded " << maxLoopCount << L" links while evaluating device " << dmDevice

That's better information to go on. But I still think you should do a "log once" type thing for a particular dmDevice to avoid needlessly filling the error log.